src/analysis/sum_integral_comparisons.lean

```diff
@@ -0,0 +1,87 @@
+/-
+Copyright (c) 2022 Kevin H. Wilson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kevin H. Wilson
+-/
+import measure_theory.measure.measure_space_def
+import measure_theory.integral.interval_integral
+import algebra.order.floor
+import analysis.special_functions.integrals
+
+/-!
+# Comparing sums and integrals
+
+## Summary
+
+It is often the case that error terms in analysis can be computed by converting the
+an infinite sum to the improper integral of an antitone function. This file enables that.
+
+## Main Results
+
+* `antitone_integral_le_sum`: The integral of an antitone function is at most the sum of its values
+  integer points (left hand side)
+
+## Tags
+
+analysis, comparison, asymptotics
+-/
+
+open_locale big_operators
+
+lemma convert_finite_sum_to_interval_integral {m n : ℕ} {f : ℝ → ℝ} (hmn : m ≤ n) :
+  ∑ (i : ℕ) in finset.Ico m n, ∫ (x : ℝ) in ↑i..↑i + 1, f ↑i =
+  ∫ (x : ℝ) in m..n, f ↑⌊x⌋₊ :=
+begin
+  conv
+  { to_lhs, congr, skip, funext, rw integral_const_on_unit_interval, },
+  obtain ⟨n', hn'⟩ := le_iff_exists_add.mp hmn,
+  rw [hn', ←interval_integral.sum_integral_adjacent_intervals'],
+  { apply finset.sum_congr rfl,
+    intros x hx,
+    rw ← @integral_const_on_unit_interval x (f ↑x),
+    apply interval_integral.integral_congr_ae,
+    rw [set.interval_oc_of_le (calc (x : ℝ) ≤ x + 1 : by simp), measure_theory.ae_iff],
+    refine real.measure_space.volume.to_outer_measure.mono_null _
+      (@real.volume_singleton ((x : ℝ) + 1)),
+    simp only [set.mem_Ioc, and_imp, not_forall, exists_prop, set.subset_singleton_iff,
+      set.mem_set_of_eq],
+    intros y hy hy' hf,
+    cases lt_or_eq_of_le hy',
+    { exfalso, rw nat.floor_eq_on_Ico x y ⟨le_of_lt hy, h⟩ at hf, exact hf rfl, },
+    { exact h, }, },
+  { intros k hk,
+    rw [interval_integrable_iff, set.interval_oc_of_le (calc (k : ℝ) ≤ ↑(k + 1) : by simp)],
+    have : measure_theory.integrable_on (λ x : ℝ, f ↑k) (set.Ioc (k : ℝ) (↑k + 1))
+      real.measure_space.volume,
+    { simp, },
+    apply this.congr_fun',
+    rw [filter.eventually_eq, measure_theory.ae_iff],
+    simp only [measure_theory.measure.restrict_apply', measurable_set_Ioc],
+    refine measure_theory.outer_measure.mono_null _ _ (@real.volume_singleton ((k : ℝ) + 1)),
+    simp only [set.subset_singleton_iff, set.mem_inter_eq, set.mem_set_of_eq, set.mem_Ioc, and_imp],
+    intros y hf hy hy',
+    cases lt_or_eq_of_le hy',
+    { exfalso, rw k.floor_eq_on_Ico y ⟨le_of_lt hy, h⟩ at hf, exact hf rfl, },
+    { exact h, }, },
+end
+
+lemma antitone_integral_le_sum {a b : ℕ} {f : ℝ → ℝ} (hab : a ≤ b)
+  (hf : antitone_on f (set.Icc a b)) : ∫ x in a..b, f x ≤ ∑ x in finset.Ico a b, f x :=
```

Comment:
I'm not completely satisfied by this statement, as it requires both bounds to be integers. I think I'd rather have a statement of the form
```
lemma antitone_on.integral_le_sum {x₀ : ℝ} {a : ℕ} {f : ℝ → ℝ}
  (hf : antitone_on f (set.Icc x₀ (x₀ + a))) :
  ∫ x in x₀..(x₀ + a), f x ≤ ∑ i in finset.range a, f (x₀ + i) :=
```
(from which your statement readily follows -- and you could even have it as a special case with its own lemma).

Note that I have switched to dot notation for the name of the lemma, to make it easier to use.

Reply:
I admit that I'm not a big fan of the `range` lemmas as it's extremely common to need short sums or dyadic sums when doing actual analysis and so there's a bunch of yoga around changing the bounds around.

However, perhaps this is obviated by a short tutorial on "here's how you should consider writing an analysis proof with dyadic sums" or somesuch whenever that comes around.

Anyway, happy to change it!

Reply:
OK, I made this change, but I'm at a weird place where I can't get the unifier to unify and I can't get `conv` to let me access the bounds of the interval integral. I've left it as a `sorry` for now. Would you happen to know the trick I need here?

Thanks!

Reply:
OK, I have figured out a version of this. See the new commit.